model: add Janus Pro for image understanding

ravenouse opened the pull request:
This pull request introduces support for the Janus‑Pro 1B and Janus‑Pro 7B models within the llama.cpp framework. 

The focus of this update is on image understanding (i.e., visual-input → textual or conceptual output). 
Image generation is not covered by this PR.

## Usage & Current Progress
**Convert models to GGUF files**:
```
# Convert the base Janus-Pro 1B model
python convert_hf_to_gguf.py deepseek-community/Janus-Pro-1B \
    --remote \
    --outfile janus-pro-1b-f16.gguf \
    --outtype f16

# Convert the mmproj component
python convert_hf_to_gguf.py deepseek-community/Janus-Pro-1B \
    --remote \
    --outfile mmproj-janus-pro-1b-f16.gguf \
    --outtype f16 \
    --mmproj
```

**The converted GGUF files can be accessed here**: https://huggingface.co/Ericwang/Janus-Pro-1B-GGUF

**Run the model**:
```
# Build the project:
cmake -B build
cmake --build build --target llama-mtmd-cli

./build/bin/llama-mtmd-cli \
    -m janus-pro-1b-f16.gguf \
    --mmproj mmproj-janus-pro-1b-f16.gguf \
    --chat-template deepseek
```

## References
Janus-Pro 1B model card:
https://huggingface.co/deepseek-community/Janus-Pro-1B

Janus-Pro 7B model card:
https://huggingface.co/deepseek-community/Janus-Pro-7B

Configurations:
https://huggingface.co/deepseek-community/Janus-Pro-1B/blob/main/config.json
https://huggingface.co/deepseek-community/Janus-Pro-7B/blob/main/config.json

HF Implementation:
https://github.com/huggingface/transformers/tree/main/src/transformers/models/janus

ravenouse committed:
Add support for Janus Pro

ravenouse commented:
Tested with this image: https://www.pixelstalk.net/wp-content/uploads/2016/04/Golden-retriever-dogs-high-definition-wallpapers.jpg

<img width="1653" height="504" alt="Screenshot 2025-10-31 at 4 02 00 PM" src="https://github.com/user-attachments/assets/15e4c465-2d7c-411b-8157-6f749b31784a" />

CISC commented on gguf-py/gguf/tensor_mapping.py:
```diff
@@ -1193,36 +1165,27 @@ class TensorNameMap:
             "model.vision_tower.embeddings.patch_embeddings.projection", # Intern-S1
             "vpm.embeddings.patch_embedding",
-            "model.vision_model.embeddings.patch_embedding", # SmolVLM
+            "model.vision_model.embeddings.patch_embedding", # SmolVLM, Janus Pro
```
```suggestion
            "model.vision_model.embeddings.patch_embedding", # SmolVLM
```
No need to stack on for every model, the list will get very long that way. :)

CISC commented on gguf-py/gguf/tensor_mapping.py:
```diff
@@ -1193,36 +1165,27 @@ class TensorNameMap:
             "model.vision_tower.embeddings.position_embeddings", # Intern-S1
             "vpm.embeddings.position_embedding",
-            "model.vision_model.embeddings.position_embedding", # SmolVLM
+            "model.vision_model.embeddings.position_embedding", # SmolVLM, Janus Pro
```
```suggestion
            "model.vision_model.embeddings.position_embedding", # SmolVLM
```

ngxson commented on tools/mtmd/clip.cpp:
```diff
@@ -4790,8 +4498,8 @@ int clip_n_mmproj_embd(const struct clip_ctx * ctx) {
             return ctx->model.mm_2_w->ne[1];
-        case PROJECTOR_TYPE_COGVLM:
-            return ctx->model.mm_4h_to_h_w->ne[1];
+        case PROJECTOR_TYPE_JANUS_PRO:
```
Maybe better to move this after `case PROJECTOR_TYPE_QWEN25VL` ?

We can read the shape of bias tensor instead of the weight, they should be the same

ravenouse committed:
Update gguf-py/gguf/tensor_mapping.py

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

ravenouse committed:
Update gguf-py/gguf/tensor_mapping.py

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

ravenouse committed:
Address reviewer suggestions

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

ravenouse committed:
Add JANUS_PRO constant

ravenouse committed:
Update clip model handling

Co-authored-by: Xuan-Son Nguyen <son@huggingface.co>

ngxson commented on tools/mtmd/clip.cpp:
```diff
@@ -1715,6 +1509,34 @@ struct clip_graph {
+        ggml_build_forward_expand(gf, cur);
+
+        return gf;
+    }
```
I think we can now merge this function with `build_siglip()`, you only need to add another `else if (ctx->proj_type() == ...)`

ngxson commented on tools/mtmd/clip.cpp:
```diff
@@ -3119,16 +2809,12 @@ struct clip_model_loader {
+                    model.mm_0_w = get_tensor(string_format(TN_LLAVA_PROJ, 0, "weight"));
+                    model.mm_0_b = get_tensor(string_format(TN_LLAVA_PROJ, 0, "bias"), false);
+                    model.mm_1_w = get_tensor(string_format(TN_LLAVA_PROJ, 1, "weight"));
+                    model.mm_1_b = get_tensor(string_format(TN_LLAVA_PROJ, 1, "bias"), false);
```
Does this work?

```suggestion
                    model.mm_0_b = get_tensor(string_format(TN_LLAVA_PROJ, 0, "bias"));
                    model.mm_1_w = get_tensor(string_format(TN_LLAVA_PROJ, 1, "weight"));
                    model.mm_1_b = get_tensor(string_format(TN_LLAVA_PROJ, 1, "bias"));
```

ravenouse committed:
Update tools/mtmd/clip.cpp

Co-authored-by: Xuan-Son Nguyen <thichthat@gmail.com>

ravenouse committed:
Refactor JANUS_PRO handling in clip.cpp

Co-authored-by: Xuan-Son Nguyen <son@huggingface.co>

ravenouse commented:
Thanks again for the review!

Just updated the code and tested it again. 

I've updated the code and retested it with the following image:
https://1.bp.blogspot.com/-tLB0HRLcOp4/Tj4Pvhsq6vI/AAAAAAAAAG8/h6ahy6g4GJI/s1600/Llama_lying_down.jpg

<img width="1275" height="249" alt="Screenshot 2025-11-02 at 10 17 44 AM" src="https://github.com/user-attachments/assets/dbde31c4-c1f8-4be3-bdf8-4c7da332d98b" />

\
PS: The forced push was made to correct a formatting typo in the commit message.

ngxson committed:
Merge branch 'master' into januspro

ngxson reviewed (approved):
Looking good! Merging once the CI passes

CISC commented:
Fix all the whitespace errors though. :)
https://github.com/ggml-org/llama.cpp/actions/runs/19016845485/job/54305935425?pr=16906

ngxson committed:
Update tools/mtmd/clip.cpp

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

ngxson committed:
em whitespace
